src/Language/Haskell/TH/Syntax/Compat.hs

```diff
@@ -341,3 +388,162 @@ instance Quote m => Quasi (QuoteToQuasi m) where
 #if MIN_VERSION_template_haskell(2,16,0)
   qReifyType          = qtqError "qReifyType"
 #endif
+
+-------------------------------------------------------------------------------
+-- Code
+-------------------------------------------------------------------------------
+
+-- $code
+-- The 'Code' type (first proposed in
+-- <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0195-code-texp.rst GHC Proposal 195>)
+-- was introduced in @template-haskell-2.17.0.0@. This module defines a version
+-- of 'Code' that is backward-compatible with older @template-haskell@
+-- releases and is forward-compatible with the existing 'Code' class.
+-- In addition to 'Code', this module also backports the functions in
+-- "Language.Haskell.TH.Syntax" that manipulate 'Code' values.
+--
+-- One troublesome aspect of writing backwards-compatible code involving 'Code'
+-- is that GHC 9.0 changed the types of typed Template Haskell splices. Before,
+-- they were of type @'Q' ('TExp' a)@, but they are now of type @'Code' 'Q' a@.
+-- The 'IsCode' class can be used to paper over the difference between these
+-- two types. For more details, consult the Haddocks for 'IsCode'.
+
+#if MIN_VERSION_template_haskell(2,9,0)
+-- | A class that allows one to smooth over the differences between
+-- @'Code' 'm' a@ (the type of typed Template Haskell quotations on
+-- @template-haskell-2.17.0.0@ or later) and @'m' ('TExp' a)@ (the type of
+-- typed Template Haskell quotations on older versions of @template-haskell@).
+-- Here are two examples that demonstrate how to use each method of 'IsCode':
+--
+-- @
+-- &#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;
+--
+-- import "Language.Haskell.TH"
+-- import "Language.Haskell.TH.Syntax.Compat"
+--
+-- -- 'toCode' will ensure that the end result is a 'Code', regardless of
+-- -- whether the quote itself returns a 'Code' or a 'TExp'.
+-- myCode :: 'Code' 'Q' Int
+-- myCode = 'toCode' [|| 42 ||]
+--
+-- -- 'fromCode' will ensure that the input 'Code' is suitable for splicing
+-- (i.e., it will return a 'Code' or a 'TExp' depending on the
+-- @template-haskell@ version in use).
+-- fortyTwo :: Int
+-- fortyTwo = $$('fromCode' myCode)
+-- @
+class IsCode q a c | c -> a q where
+  -- | Convert something to a 'Code'.
+  toCode   :: c -> Code q a
+  -- | Convert to something from a 'Code'.
+  fromCode :: Code q a -> c
```

Comment:
These are good. 

Havng `SpliceQ` so one could write type signatures like

```haskell
mySplice :: SpliceQ Int
mySplice = ...

-- and elsewhere
value = $$(mySplice)
``` 

That would allow library writers to hide details in the library, at least in some cases.

Reply:
I assume that you would want `SpliceQ` to be defined like this?

```hs
#if MIN_VERSION_template_haskell(2,17,0)
type SpliceQ = CodeQ
-- More generally,
-- type Splice = Code
-- type SpliceQ = Splice Q
#else
type SpliceQ = TExpQ
-- More generally,
-- type Splice m a = m (TExp a)
-- type SpliceQ a = Splice Q a
#endif
```

If so, I'm somewhat conflicted about this, since this would introduce an API that uses completely different types depending on the `template-haskell` version. My gut reaction would be to suggest that everyone first convert to `Code` (the modern API type) using `toCode` and then simply manipulate `Code` values.

On the other hand, I will admit that I have a use case for `Splice` myself: it could useful for implementing `liftTyped`, whose return type changed in version 2.17. That is, the type of `liftTyped` can be thought of as `(Lift a, Quote q) => q Exp -> Splice q a`. Perhaps I will add `Splice`/`SpliceQ` and include a disclaimer that you should only use them if you know what you are doing.